pkg/operator/controller.go

```diff
@@ -0,0 +1,518 @@
+	appsset "k8s.io/client-go/kubernetes/typed/apps/v1"
+	coreset "k8s.io/client-go/kubernetes/typed/core/v1"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+
+	configapiv1 "github.com/openshift/api/config/v1"
+	configset "github.com/openshift/client-go/config/clientset/versioned"
+	configsetv1 "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	configinformers "github.com/openshift/client-go/config/informers/externalversions"
+
+	tunedv1 "github.com/openshift/cluster-node-tuning-operator/pkg/apis/tuned/v1"
+	ntoclient "github.com/openshift/cluster-node-tuning-operator/pkg/client"
+	ntoconfig "github.com/openshift/cluster-node-tuning-operator/pkg/config"
+	tunedset "github.com/openshift/cluster-node-tuning-operator/pkg/generated/clientset/versioned"
+	tunedinformers "github.com/openshift/cluster-node-tuning-operator/pkg/generated/informers/externalversions"
+	ntomf "github.com/openshift/cluster-node-tuning-operator/pkg/manifests"
+)
+
+const (
+	// workqueue related constants
+	wqKindPod             = "pod"
+	wqKindNode            = "node"
+	wqKindClusterOperator = "clusteroperator"
+	wqKindConfigMap       = "configmap"
+	wqKindDaemonSet       = "daemonset"
+	wqKindTuned           = "tuned"
+	wqKindProfile         = "profile"
+)
+
+// Controller is the controller implementation for Tuned resources
+type Controller struct {
+	kubeconfig *restclient.Config
+
+	// workqueue is a rate limited work queue. This is used to queue work to be
+	// processed instead of performing it as soon as a change happens.
+	workqueue workqueue.RateLimitingInterface
+
+	listers *ntoclient.Listers
+	clients *ntoclient.Clients
+
+	pc *ProfileCalculator
+}
+
+func NewController() (*Controller, error) {
+	kubeconfig, err := ntoclient.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	listers := &ntoclient.Listers{}
+	clients := &ntoclient.Clients{}
+	controller := &Controller{
+		kubeconfig: kubeconfig,
+		workqueue:  workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		listers:    listers,
+		clients:    clients,
+		pc:         NewProfileCalculator(listers, clients),
+	}
+
+	// Initial event to bootstrap CR if it doesn't exist.
+	controller.workqueue.AddRateLimited(wqKey{kind: wqKindTuned, name: tunedv1.TunedDefaultResourceName})
+
+	return controller, nil
+}
+
+type wqKey struct {
+	kind      string // object kind
+	namespace string // object namespace
+	name      string // object name
+	event     string // object event type (add/update/delete) or pass the full object on delete
+}
+
+// eventProcessor is a long-running function that will continually
+// read and process a message on the workqueue.
+func (c *Controller) eventProcessor() {
+	for {
+		// Wait until there is a new item in the working queue
+		obj, shutdown := c.workqueue.Get()
+		if shutdown {
+			return
+		}
+
+		klog.V(2).Infof("get event from workqueue")
```

Comment:
Only small logging change, should be "got event"

Reply:
Thanks, fixed!